Crossfade smoothers

paulfd opened the pull request:
This adds a crossfade smoother, regarding previous reports in #153 #22 #19 . There was an improvement with the global gain smoothing in #181 which was reverted in #309 because applying it globally on the gain had undesirable side-effects on e.g. drums with big transients.

This PR:
- Adds a "shortcut" in all smoothers. A smoother will just copy the span if i) the input span to smooth seems to be basically constant, and ii) the first value of the span is the same as the current state of the smoother.
- Add a separate process for the crossfades.

I'll evaluate the performance loss/gain.

@jpcima When doing this modulation matrix I think it'd be good to find a clean way to shortcut as much as possible. Let's discuss this further 🙂

paulfd committed:
Add short paths for the smoothers
When the input is basically constant in particular

paulfd committed:
Smooth the crossfades

paulfd committed:
Further separate the crossfades and initialize them properly

paulfd commented:
It seems to work OK, and does not impact performance too much.

https://box.ferrand.cc/sfizz-reports/2020-07-06-xfade-history.html

Showcasing the perf-tracking work in progress :) The 3 big hits in performance in recent history were due to:
1. The voice stealing algorithm that uses an envelope follower on every voice.
2. The resampler going from linear to polynomial by default
3. The global smoother

Item 2. is not really negotiable and can be user-deactivated, item 3. was improved by deactivating the global smoother in #309 and the current PR recovers the functionality of the global gain smoother specifically on crossfades for a low overhead. For item 1. I'm using a lowpass filter as an envelope follower, but @jpcima mentioned that Faust had envelope followers so we might test those to try and reduce the perf hit at some point.

jpcima commented on src/sfizz/Smoothers.cpp:
```diff
@@ -27,10 +27,25 @@ void Smoother::reset(float value)
+    const auto midValue = input[input.size() / 2];
+    const bool shortcut = (
+        input.front() == input.back()
+        && input.front() == midValue
```
how valid is this as a guarantee that signal may be constant?
if we are smoothing a CC, it may well happen as a coincidence that 3 points of the CC stream are equal but the rest of signal isn't.

- beside even if the condition for shortcut is valid, what if we leave a state of filter smoothing which has not completed? do we take a discontinuity?
- same after we leave shortcut state to go back to the filter, we accept the discontinuity there also?

paulfd replied:
> how valid is this as a guarantee that signal may be constant?
> if we are smoothing a CC, it may well happen as a coincidence that 3 points of the CC stream are equal but the rest of signal isn't.

For sure we may miss these cases.

> 
>     * beside even if the condition for shortcut is valid, what if we leave a state of filter smoothing which has not completed? do we take a discontinuity?
> 
>     * same after we leave shortcut state to go back to the filter, we accept the discontinuity there also?

I think for these the condition
```c++
&& input.front() == current()
```
will be false.

paulfd replied:
Maybe another way could be to check if there was an event in the block. If no event AND if the filter has finished smoothing to the expected value, then we can shortcut.

jpcima replied:
Does this mean
- if there is no event ever received,the code is fast
- the moment the first cc arrives, the code is slow?

what is going to happen with floating point IIR? is a filter going to converge moderately "fast" to equal the target value? will it be guaranteed to converge ever?

jpcima replied:
Anyway, I think shortcut can be deduced according to these conditions
- the cc stream is stationary ; defined as having no events in the current cycle
- the filter output is within some reasonable epsilon of the target

paulfd replied:
> Anyway, I think shortcut can be deduced according to these conditions
> 
>     * the cc stream is stationary ; defined as having no events in the current cycle
> 
>     * the filter output is within some reasonable epsilon of the target

This is basically what I meant in the last reply. Let me try to provide an implementation.

paulfd replied:
New tentative in dda4a94

paulfd committed:
The process call in the smoother now takes an external canShortcut parameter
If the filter can shortcut, it will check that the first input of the span is within a reasonable range of the current value of the filter, and if so it resets it to this value and proceeds as if it were not smoothing

paulfd commented:
Something weird is happening to the macOS build.

```
==> Upgrading python 3.7.5 -> 3.7.8 

==> Downloading https://homebrew.bintray.com/bottles/python-3.7.8.high_sierra.bottle.tar.gz

==> Downloading from https://akamai.bintray.com/d2/d2315dbe4f3266b7c6efe8bfda3ae0bedc60792cfff00f4bcdf87e5923274ef3?__gda__=exp=1594117301~hmac=5454a7cc1c5f8a97d7c615f34bd1fb566b63bff9eaf0b91d914715babab3137f&response-content-disposition=attachment%3Bfilename%3D%22python-3.7.8.high_sierra.bottle.tar.gz%22&response-content-type=application%2Fgzip&requestInfo=U2FsdGVkX18jMNrNRc2Amfoydar178dCFXVjHn4PCs5tb-5Zi6BHFBG-4_v6BG8-GEJxLJ9S3JRf8jd-bUyN2HpHqMo3ZZOxMML8Uo7DB1JSNVSbZVwZClASKk0qetNczsybzSSDOcx6JZk95NA9rQ&response-X-Checksum-Sha1=22e6d82b35f46437795c84e371b7b20c772a3fe6&response-X-Checksum-Sha2=d2315dbe4f3266b7c6efe8bfda3ae0bedc60792cfff00f4bcdf87e5923274ef3

==> Pouring python-3.7.8.high_sierra.bottle.tar.gz

Error: The `brew link` step did not complete successfully

The formula built, but is not symlinked into /usr/local

Could not symlink Frameworks/Python.framework/Headers

Target /usr/local/Frameworks/Python.framework/Headers

is a symlink belonging to python@2. You can unlink it:

  brew unlink python@2

To force the link and overwrite all conflicting files:

  brew link --overwrite python

To list all files that would be deleted:

  brew link --overwrite --dry-run python

Possible conflicting files are:

/usr/local/Frameworks/Python.framework/Headers -> /usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Headers

/usr/local/Frameworks/Python.framework/Python -> /usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Python

/usr/local/Frameworks/Python.framework/Resources -> /usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Resources

/usr/local/Frameworks/Python.framework/Versions/Current -> /usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/Current

==> /usr/local/Cellar/python/3.7.8/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.8/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt

==> /usr/local/Cellar/python/3.7.8/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.8/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt

==> /usr/local/Cellar/python/3.7.8/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python/3.7.8/bin --install-lib=/usr/local/lib/python3.7/site-packages --single-version-externally-managed --record=installed.txt

==> Caveats

Python has been installed as

  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to

`python3`, `python3-config`, `pip3` etc., respectively, have been installed into

  /usr/local/opt/python/libexec/bin

You can install Python packages with

  pip3 install <package>

They will install into the site-package directory

  /usr/local/lib/python3.7/site-packages

See: https://docs.brew.sh/Homebrew-and-Python

==> Summary

🍺  /usr/local/Cellar/python/3.7.8: 4,007 files, 61.1MB

Removing: /usr/local/Cellar/python/3.7.5... (4,049 files, 61.9MB)
```

It's seems that there are no real errors in fact?

jpcima commented:
Not familiar to me, there is a homebrew issue
https://github.com/Homebrew/brew/issues/1742
and a link to a fix commit at the bottom, we may be able to use the same

paulfd committed:
Trying to solve the  macos CI problem
